File: Problem_Solving_Paradigms/Dynamic_Programming/UVa_216_Getting_in_Line.py
# https://onlinejudge.org/external/2/216.pdf

from itertools import permutations
import sys


def distance(p1, p2):
    return (abs(p1[0] - p2[0]) ** 2 + abs(p1[1] - p2[1]) ** 2) ** 0.5


def solution(n, coords):
    graph = [[0] * n for _ in range(n)]
    min_perm, min_dist = None, 1e9

    for i in range(n):
        for j in range(i + 1, n):
            graph[i][j] = graph[j][i] = distance(coords[i], coords[j]) + 16

    for perm in permutations(range(n)):
        dist = sum(graph[perm[k]][perm[k + 1]] for k in range(n - 1))
        if dist < min_dist:
            min_dist = dist
            min_perm = perm
    for i in range(1, n):
        s, e = min_perm[i - 1], min_perm[i]
        x1, y1 = coords[s]
        x2, y2 = coords[e]
        print("Cable requirement to connect ({},{}) to ({},{}) is {:.2f} feet.".format(
            x1, y1, x2, y2, graph[s][e]))
    print("Number of feet of cable required is {:.2f}.".format(min_dist))


inp = iter(sys.stdin.readlines())
count = 1
while True:
    n = int(next(inp))
    if not n:
        break
    coords = [list(map(int, (next(inp).split()))) for _ in range(n)]
    print("**********************************************************")
    print(f"Network #{count}")
    count += 1
    solution(n, coords)

File: UVa/Problem Solving/DP/UVa_216_Getting in Line.cpp
// https://onlinejudge.org/external/2/216.pdf

#include <bits/stdc++.h>
using namespace std;
#define REP(i, n) for (int i = 0; i < n; ++i)

typedef pair<int, int> ii;

int n;
ii computers[8];
double dist[8][8];

double distance(ii a, ii b)
{
    return sqrt(pow(a.first - b.first, 2) + pow(a.second - b.second, 2)) + 16.0;
}

int main()
{
    int tc = 1;
    while (scanf("%d", &n), n)
    {
        REP(i, n)
        {
            scanf("%d %d", &computers[i].first, &computers[i].second);
        }

        REP(i, n)
        {
            REP(j, i)
            {
                double d = distance(computers[i], computers[j]);
                dist[i][j] = d;
                dist[j][i] = d;
            }
        }
        vector<int> network;
        vector<int> computer_idx;
        double min_cable = numeric_limits<double>::max();

        REP(i, n)
        {
            computer_idx.push_back(i);
        }

        do
        {
            double cable = 0;
            REP(i, n - 1)
            {
                cable += dist[computer_idx[i]][computer_idx[i + 1]];
            }

            if (cable < min_cable)
            {
                min_cable = cable;
                network = computer_idx;
            }

        } while (next_permutation(computer_idx.begin(), computer_idx.end()));

        printf("**********************************************************\n");
        printf("Network #%d\n", tc++);
        REP(i, n - 1)
        {
            printf("Cable requirement to connect (%d,%d) to (%d,%d) is %.2f feet.\n",
                   computers[network[i]].first, computers[network[i]].second,
                   computers[network[i + 1]].first, computers[network[i + 1]].second,
                   dist[network[i]][network[i + 1]]);
        }
        printf("Number of feet of cable required is %.2f.\n", min_cable);
    }

    return 0;
}